site/public/js/submission-page.js

```diff
@@ -7,3 +8,124 @@ function openActionsPopup(popup_css, element_id) {
     my_window.document.close();
     my_window.focus();
 }
+
+let days, hours, mins, seconds = 0;
+let curTime = 0;
+let lastTime = 0;
+let deadline = 0;
+let user_deadline = 0;
+let startTime = 0;
+let width = 0;
+let allowedTime = 0;
+let gradeable_id = '';
+let ticks_till_update = 600000;
+
+function initializeTimer(gradeableID) {
+    gradeable_id = gradeableID;
+    syncWithServer();
+}
+
+function syncWithServer() {
+    // eslint-disable-next-line no-undef
+    const url = buildCourseUrl(['gradeable', gradeable_id, 'time_remaining_data']);
+    $.ajax({
+        url,
+        type: 'GET',
+        processData: false,
+        contentType: false,
+        success: function(res) {
+            lastTime = Date.now();
+            const response = JSON.parse(res);
+            ticks_till_update = 600000;
+            if (response.status === 'success') {
+                const { data } = response;
+                if (Object.prototype.hasOwnProperty.call(data, 'user_allowed_time_deadline')) {
+                    user_deadline = data.user_allowed_time_deadline;
+                }
+                if (Object.prototype.hasOwnProperty.call(data, 'user_allowed_time')) {
+                    allowedTime = data.user_allowed_time;
+                }
+                if (Object.prototype.hasOwnProperty.call(data, 'user_start_time')) {
+                    startTime = data.user_start_time;
+                }
+                curTime = data.current_time;
+                deadline = data.deadline;
+                updateTime();
+            }
+            else {
+                // eslint-disable-next-line no-undef
+                displayErrorMessage('Something went wrong while starting the timer');
+            }
+        },
+        error: function (err) {
```

Comment:
What should be the expected behavior if the server cannot be contacted?  Should the progress bar continue moving forward?  Should the progress bar disappear?  Should some sort of message appear indicating that the user is now offline/unable to communicate with the server?  I'm personally in favor of the progress bar continuing based upon the system time and then displaying some sort of error message saying that the server can't be contacted.  This is also useful for other things such as the user turning off their internet or the server going down during a timed test.

Reply:
Right I forgot to finish the error handling part. I think I will add an argument into syncWithServer which will determine if this is a necessary call or not. Necessary means the first time its called or if the time was set off by more than 5 seconds. In this case I will have the timer error out since it could be off by a lot at this point or not even work. A non necessary call would just be the every 10 minute check which I would just continue counting. Once they connect to the network it will continue to work as normal.